src/utils/theme.js

```diff
@@ -0,0 +1,42 @@
+import React from 'react'
+import _ from 'lodash/fp'
+import F from 'futil-js'
+
+export let ThemeContext = React.createContext()
+
+let hasNested = _.curry(
+  (key, theme) =>
+    key && theme && F.findIndexed((v, k) => _.startsWith(`${key}.`, k), theme)
+)
+
+export let mergeNestedTheme = (theme, key) =>
+  F.when(
+    hasNested(key),
+    _.flow(
+      _.pickBy((val, k) => _.startsWith(`${key}.`, k)),
+      _.mapKeys(x => _.replace(`${key}.`, '', x)),
+      _.defaults(theme)
+    )
+  )(theme)
+
+export let mergeNestedThemePath = (theme, path) =>
+  _.reduce((acc, key) => mergeNestedTheme(acc, key), theme, path || [])
+
+export let ThemeConsumer = ({ children, path }) => (
+  <ThemeContext.Consumer>
+    {theme => children(mergeNestedThemePath(theme, path))}
+  </ThemeContext.Consumer>
+)
+
+export let withTheme = name => Component => ({
+  theme: propTheme,
+  ...props
+}) => {
+  let contextTheme = mergeNestedTheme(React.useContext(ThemeContext), name)
+  let newTheme = _.merge(contextTheme, propTheme)
+  return (
+    <ThemeContext.Provider value={newTheme}>
```

Comment:
I just realized - this will also merge the prop theme onto the context for children, which means component theme stuff will cascade down to children, too. I wonder if that's what we want 🤔

Reply:
It was intentional to make the last story work (merging nested theme props) - so tentatively yes